src/creatures/npcs/npc.cpp

```diff
@@ -590,16 +613,27 @@ void Npc::setPlayerInteraction(uint32_t playerId, uint16_t topicId /*= 0*/) {
 		return;
 	}
 
-	turnToCreature(creature);
+	if (playerInteractionsOrder.empty() || std::ranges::find(playerInteractionsOrder, playerId) == playerInteractionsOrder.end()) {
+		playerInteractionsOrder.emplace_back(playerId);
+		turnToCreature(creature);
+	}
 
 	playerInteractions[playerId] = topicId;
 }
 
 void Npc::removePlayerInteraction(std::shared_ptr<Player> player) {
+	auto view = std::ranges::remove(playerInteractionsOrder, player->getID());
+	playerInteractionsOrder.erase(view.begin(), view.end());
 	if (playerInteractions.contains(player->getID())) {
 		playerInteractions.erase(player->getID());
 		player->closeShopWindow();
 	}
+
+	if (!playerInteractionsOrder.empty()) {
+		if (const auto creature = g_game().getCreatureByID(playerInteractionsOrder.back()); creature) {
```

Comment:
Set as const reference will avoid increasing the reference counter, and the if already checks if the ptr is empty, there is no need to use "; creature"
```suggestion
		if (const auto &creature = g_game().getCreatureByID(playerInteractionsOrder.back())) {
```

Reply:
> Set as const reference will avoid increasing the reference counter, and the if already checks if the ptr is empty, there is no need to use "; creature"

Fixed